### ISSUE TITLE
meta: tor uplift: privacy.resistFingerprinting

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [arkenfox/user.js#7]
Locked sticky issue for tracking privacy.resistFingerprinting
 - [Tor Uplift](https://wiki.mozilla.org/Security/Tor_Uplift/Tracking)
 - [MVP Tracking](https://wiki.mozilla.org/Security/Fingerprinting)
 - [Bugzilla META RFP ticket](https://bugzilla.mozilla.org/show_bug.cgi?id=1329996)
 - [Tor Uplift Dependency Tree](https://bugzilla.mozilla.org/showdependencytree.cgi?id=1260929&hide_resolved=0)
 - [Mozilla Support: Firefox's protection against fingerprinting](https://support.mozilla.org/en-US/kb/firefox-protection-against-fingerprinting)

Last updated: see changelog at foot (regularly cleaned out)

### RESOLVED & ADDED TO USER.JS <sup>if applicable</sup>

:small_red_triangle_down: FF41
 * [418986](https://bugzilla.mozilla.org/show_bug.cgi?id=418986) enable privacy.resistFingerprinting (limit window.screen & CSS media etc)
   // user_pref("privacy.resistFingerprinting", true);

:small_red_triangle_down: FF50
 * [1281949](https://bugzilla.mozilla.org/show_bug.cgi?id=1281949) spoof screen orientation
 * [1281963](https://bugzilla.mozilla.org/show_bug.cgi?id=1281963) hide the contents of navigator.plugins and navigator.mimeTypes
   * FF53: [1324044](https://bugzilla.mozilla.org/show_bug.cgi?id=1324044) fixes GetSupportedNames in nsMimeTypeArray and nsPluginArray

:small_red_triangle_down: FF55
 * [1330890](https://bugzilla.mozilla.org/show_bug.cgi?id=1330890) spoof timezone as UTC 0
   * FF58: [818634](https://bugzilla.mozilla.org/show_bug.cgi?id=818634) deprecates Date.toLocaleFormat
   * FF60: [1409973](https://bugzilla.mozilla.org/show_bug.cgi?id=1409973) fixes Date.toLocaleDateString and Intl.DateTimeFormat
 * [1330882](https://bugzilla.mozilla.org/show_bug.cgi?id=1330882) new window sizes to round to hundreds & override prefs
   // user_pref("privacy.window.maxInnerWidth", 1600);
   // user_pref("privacy.window.maxInnerHeight", 900);
 * [1360039](https://bugzilla.mozilla.org/show_bug.cgi?id=1360039) spoof navigator.hardwareConcurrency as 2
 * [1217238](https://bugzilla.mozilla.org/show_bug.cgi?id=1217238) reduce precision of time exposed by javascript

:small_red_triangle_down: FF56
* [1369303](https://bugzilla.mozilla.org/show_bug.cgi?id=1369303) spoof/disable performance API
* [1333651](https://bugzilla.mozilla.org/show_bug.cgi?id=1333651) & [1383495](https://bugzilla.mozilla.org/show_bug.cgi?id=1383495) & [1396468](https://bugzilla.mozilla.org/show_bug.cgi?id=1396468) spoof navigator API
   * FF57: [1393283](https://bugzilla.mozilla.org/show_bug.cgi?id=1393283) spoof as ESR instead of 10s
   * FF59: [1415488](https://bugzilla.mozilla.org/show_bug.cgi?id=1415488) UA leaks over HTTP CONNECT method - fixed as a result of [1419771](https://bugzilla.mozilla.org/show_bug.cgi?id=1419771)
   * FF59: [1404608](https://bugzilla.mozilla.org/show_bug.cgi?id=1404608) do not lie about OS (limit to Windows, OSX, Android, or Linux)
   * FF59: [1418672](https://bugzilla.mozilla.org/show_bug.cgi?id=1418672) due to ESR being out of whack and Aurora/Nightly sometimes being ahead of ESR releases, the value is now temporally hardcoded to 52
   * FF60: [1418162](https://bugzilla.mozilla.org/show_bug.cgi?id=1418162) ESR algorithm fixed for 60+7's
   * FF63: [1472618](https://bugzilla.mozilla.org/show_bug.cgi?id=1472618) navigator.platform returns "Win32" (but JS UA still shows 64bit)
   * FF66: [1509829](https://bugzilla.mozilla.org/show_bug.cgi?id=1509829) upstream [Tor 26146](https://trac.torproject.org/projects/tor/ticket/26146) reduce UA HTTP headers to two OSes (Windows, Android)
   * FF67: [1511763](https://bugzilla.mozilla.org/show_bug.cgi?id=1511763) ESR cadence fixed to 60+8's
   * FF68: [1511434](https://bugzilla.mozilla.org/show_bug.cgi?id=1511434) UA Spoof -> Windows 10, OS 10.14, Android 8.1
   * FF78: [1599188](https://bugzilla.mozilla.org/show_bug.cgi?id=1599188) version algorithm hardcoded in 78+ to cadence 13 
   * FF78: [1635011](https://bugzilla.mozilla.org/show_bug.cgi?id=1635011) UA Spoof -> OS 10.15, Android 9.0
* [1369319](https://bugzilla.mozilla.org/show_bug.cgi?id=1369319) disable device sensors
* [1369357](https://bugzilla.mozilla.org/show_bug.cgi?id=1369357) disable site specific zoom
* [1337161](https://bugzilla.mozilla.org/show_bug.cgi?id=1337161) hide gamepads from content
* [1372072](https://bugzilla.mozilla.org/show_bug.cgi?id=1372072) spoof network information API as "unknown"
* [1372069](https://bugzilla.mozilla.org/show_bug.cgi?id=1372069) block geolocation requests
   * FF63: [1441295](https://bugzilla.mozilla.org/show_bug.cgi?id=1441295) reverted: RFP no longer blocks geo
* [1333641](https://bugzilla.mozilla.org/show_bug.cgi?id=1333641) disable WebSpeech API

:small_red_triangle_down: FF57
* [1369309](https://bugzilla.mozilla.org/show_bug.cgi?id=1369309) spoof media statistics
* [1382499](https://bugzilla.mozilla.org/show_bug.cgi?id=1382499) reduce screen co-ordinate fingerprinting in Touch API
* [1217290](https://bugzilla.mozilla.org/show_bug.cgi?id=1217290) enable fingerprinting resistance for WebGL
    * FF58: [1409677](https://bugzilla.mozilla.org/show_bug.cgi?id=1409677) WebGL issues
* [1382545](https://bugzilla.mozilla.org/show_bug.cgi?id=1382545) reduce fingerprinting in Animation API
* [1354633](https://bugzilla.mozilla.org/show_bug.cgi?id=1354633) limit MediaError.message to a whitelist
* [1382533](https://bugzilla.mozilla.org/show_bug.cgi?id=1382533) enable fingerprinting resistance for Presentation API
*  [1384330](https://bugzilla.mozilla.org/show_bug.cgi?id=1384330) disable mozAddonManager Web API
   // user_pref("privacy.resistFingerprinting.block_mozAddonManager", true);

:small_red_triangle_down: FF58
 * [967895](https://bugzilla.mozilla.org/show_bug.cgi?id=967895) prompt (site permission) before allowing canvas data extraction
   * FF59: [1413780](https://bugzilla.mozilla.org/show_bug.cgi?id=1413780) when RFP=true include canvas in site permissions panel
   * FF59: [1376865](https://bugzilla.mozilla.org/show_bug.cgi?id=1376865) reduce canvas prompt fatigue by only prompting when user initiated (controlled by a temp pref `privacy.resistFingerprinting.autoDeclineNoUserInputCanvasPrompts` )
   * FF61: [1412961](https://bugzilla.mozilla.org/show_bug.cgi?id=1412961) RFP breaks extensions using canvas eg Screenshots: 
   * > until then, you can open about:debugging>Add-ons, click the extension's `manifest.url` and on that page, right click >`View Page Info`>`Permissions` and set a canvas exception
   * FF62: [1453916](https://bugzilla.mozilla.org/show_bug.cgi?id=1453916) Fixes more extension + canvas issues
 * [1397611](https://bugzilla.mozilla.org/show_bug.cgi?id=1397611) extensions can control ` privacy.resistFingerprinting`
* [1424341](https://bugzilla.mozilla.org/show_bug.cgi?id=1424341) **FYI ONLY: Timing: not added to user.js**: RFP timing & timer precision
   // privacy.reduceTimerPrecision
   // privacy.resistFingerprinting.reduceTimerPrecision.microseconds

:small_red_triangle_down: FF59
 * [1372073](https://bugzilla.mozilla.org/show_bug.cgi?id=1372073) spoof/block fingerprinting in MediaDevices API
 * [1039069](https://bugzilla.mozilla.org/show_bug.cgi?id=1039069) warn when language prefs are set to non en-US
    * `privacy.spoof_english` (pref is used internally AFAICT, we should not meddle with it)
    * at this stage the pref has not been added to the user.js, but the bugzilla has
* [1222285](https://bugzilla.mozilla.org/show_bug.cgi?id=1222285) spoof keyboard events and suppress keyboard modifier events
   * FF59.0.2: [1433592](https://bugzilla.mozilla.org/show_bug.cgi?id=1433592) don't spoof/suppress CTRL key
   * FF60: [1438795](https://bugzilla.mozilla.org/show_bug.cgi?id=1438795) fixes keydown/keyup events
   * Test: https://w3c.github.io/uievents/tools/key-event-viewer.html

:small_red_triangle_down: FF60
 * [1337157](https://bugzilla.mozilla.org/show_bug.cgi?id=1337157) disable WebGL debug renderer info (see `2011`)
* [1425462](https://bugzilla.mozilla.org/1425462) **FYI ONLY: Timing: not added to user.js**: RFP timing precision jitter
   // privacy.resistFingerprinting.reduceTimerPrecision.jitter

:small_red_triangle_down: FF62
* [1459089](https://bugzilla.mozilla.org/show_bug.cgi?id=1459089) [**Firefox for Android**] HTTP Accept-Language header no longer leaks OS locale

:small_red_triangle_down: FF63
* [1479239](https://bugzilla.mozilla.org/show_bug.cgi?id=1479239) return `no-preference` with `prefers-reduced-motion`

:small_red_triangle_down: FF64
 * [1363508](https://bugzilla.mozilla.org/show_bug.cgi?id=1363508) spoof/suppress Pointer Events

:small_red_triangle_down: FF65
* [1492766](https://bugzilla.mozilla.org/show_bug.cgi?id=1492766) pointerEvent.pointerid

:small_red_triangle_down: FF67
* [1485266](https://bugzilla.mozilla.org/show_bug.cgi?id=1485266) enforce `ui.use_standins_for_native_colors`=`true` (`2618`)
* [1407366](https://bugzilla.mozilla.org/show_bug.cgi?id=1407366) RFP letterboxing
   // user_pref("privacy.resistFingerprinting.letterboxing", true);
   // user_pref("privacy.resistFingerprinting.letterboxing.dimensions", "");
   * FF68: [1548634](https://bugzilla.mozilla.org/show_bug.cgi?id=1548634) stepped ranges
* [1485264](https://bugzilla.mozilla.org/show_bug.cgi?id=1485264) FYI: remove `dom.event.highrestimestamp.enabled`
* [1494034](https://bugzilla.mozilla.org/show_bug.cgi?id=1494034) enforce `light` with `prefers-color-scheme` - see [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme) for this standard added in FF67

:small_red_triangle_down: FF68
* [1492587](https://bugzilla.mozilla.org/show_bug.cgi?id=1492587) possible date picker locale leak fix

:small_red_triangle_down: FF70
*  [1564422](https://bugzilla.mozilla.org/show_bug.cgi?id=1564422) spoof audioContext `outputLatency` [per OS](https://hg.mozilla.org/mozilla-central/rev/c96e81ba64f3#l1.14)

:small_red_triangle_down: FF72
 * [1595823](https://bugzilla.mozilla.org/show_bug.cgi?id=1595823) spoof audioContext sampleRate

:small_red_triangle_down: FF74
 * [1607316](https://bugzilla.mozilla.org/show_bug.cgi?id=1607316) spoof pointer as coarse and hover as none on android

:small_red_triangle_down: FF78
 * [1621433](https://bugzilla.mozilla.org/show_bug.cgi?id=1621433) randomize canvas
   - note: as of writing, `isPointInPath` and `isPointInStroke` still use the white-canvas

:small_red_triangle_down: FF80
* [1653987](https://bugzilla.mozilla.org/show_bug.cgi?id=1653987) font visibility is restricted to `BaseFonts` (non Android)
   - click [here](https://searchfox.org/mozilla-central/search?path=StandardFonts*.inc)
   - the lists are hardcoded lists with two parts `kBaseFonts` and `kLangPackFonts`
   - note: this is the same as setting `layout.css.font-visibility.level` = `1`

:small_red_triangle_down: FF82
* [1461454](https://bugzilla.mozilla.org/show_bug.cgi?id=1461454) spoof smooth=true and powerEfficient=false for Supported Media in MediaCapabilities

### CHANGELOG
2020
* May 6: cleaned out changelog
* May 6: 1635011 (rfp os spoof changes) → FF78+ (under FF56 section)
* May 15: added entries for FF74 and FF78
* Jul 29: 1653987 (font visibility) → FF80+
* Nov 11: 1461454 (mediaCapabilities) → FF82+
* Dec 26: removed followups, pending, and invalid/wontfix sections since I'm not maintaining them and they're out of date

...

**Comment by Atavic** [comment]
[TheWebNeverForgets](https://github.com/citp/TheWebNeverForgets) has a paper called [Persistent Tracking Mechanisms in the Wild](https://securehomes.esat.kuleuven.be/%7Egacar/persistent/the_web_never_forgets.pdf) that explains a very underrated issue: Cookie Syncing.

**Comment by Roman-Nopantski** [comment]
Keep this to the tor uplift please. Should I create a sticky for latest off topic "news" that doesn't need an issue on its own?

**Comment by Atavic** [comment]
@crssi https://github.com/ghacksuserjs/ghacks-user.js/issues/15#issuecomment-281325415
- ##### privacy.resistFingerprinting
You can block _Clickjacking_ with uBlock: https://github.com/gorhill/uBlock/issues/892

**Comment by crssi** [comment]
As @Roman-Nopantski already noticed, I also do not find any problems at this time when TRUE.
There are no form jerks and also NoScript doesn't fires false ClearClick events.

**Comment by share3141** [comment]
OK - this may be the wrong place and wrong format but the extension "Nosquint Plus" does very weird things when user_pref("privacy.resistFingerprinting", true);.  
It shrinks the Nosquint Plus option box and sticks it in the upper left corner.  
I apologize if I am entering this issue in the wrong place - I am not educated enough to know where to put it.  I found this site while trying to figure out what the user_pref did and posted.

**Comment by Thorin-Oakenpants** [comment]
@share3141 : Where is this nosquint plus option box? Is this on your toolbar? Is it in a context menu? Are you 100% certain it is resistFingerprinting (i.e did you download a vanilla portable Firefox, add NoSquint Plus, check the problem does not exist, change the preference, and then check the problem does exist)? Can you upload a screenshot? Have you talked to the developer of the add-on?

**Comment by crssi** [comment]
If you find that `resistFingerprinting = true` is not an option for you, then you set it to false.
Better than that is to place an issue ticket of "Nosquint Plus".

**Comment by Thorin-Oakenpants** [comment]
Nah @crssi , that's not how you it should go man :)  It's more like this: I want to use `resistFingerprinting = true` but it screws up NoSquint Plus add-on. I'll go see if NoSquint Plus knows about it and if they can't fix it, then I'll move to the e10s compliant Zoom Page (note: I use this, for zoom levels - don't give a rat's backside about changing of website colors - that's what stylish is for). NoSquint Plus is also e10s compatible, for those reading this - so seems like the dev is on the ball, and is the first person you should talk to.

**Comment by crssi** [comment]
LOL... I like you cryptic humor. :)... But always need to read it 3 times.

**Comment by share3141** [comment]
I set up a new profile for FF (53b4) and installed one extension - Nosquint Plus.  I went to a web site and invoked the Nosquint option - no issues.  I set up another profile and installed Nosquint and set resistFingerprinting = true in user.js.  I opened FF and went to the same website and invoked the Nosquint option.  The option menu that pops up for Nosquint was very reduced in size and in the upper left corner of the browser.  (The Nosquint option can be invoked from the context menu.)

**Comment by Thorin-Oakenpants** [comment]
@share3141 : then you better go tell the NoSquint Plus developer - he has an email address on the AMo page. In FF52 (and 53 so far) privacy.resistFingerprinting only covers screen/windows/orientation (and mime types). The inner window doesn't lie, but other measurements do, and whatever code he is using, is falling over due to that. I might install it to actually see what happens.

**Comment by Thorin-Oakenpants** [comment]
@crssi The point of Zoom Page and NoSquint Plus (well, the squint part) is that it allows sites to remember specific zoom levels (and there are extra tweaks). The Squint name is funny, because for me, web content is usually too large - like it's primarily designed for mobiles - i.e I'm *not* squinting. I'd call my version "Honey I Shrunk the Web" and release it under the handle Rick Moranis.

**Comment by Thorin-Oakenpants** [comment]
Well, I have around 500 sites in my speed dials and I visit about 300 of them once a week, many daily. And half of those need a zoom out. So automagic... here I come.

**Comment by earthlng** [comment]
Afaik FF already remembers site-specific zoom-levels. Why do you need an add-on for that?

**Comment by crssi** [comment]
Exactly. I am missing a solution to clear site specific settings, except the "remember password" choice because of annoying balloon... oh and obviously the "zoom" part.

**Comment by Thorin-Oakenpants** [comment]
It doesn't remember chrome:// zoom levels and my speed dials are at 75% so I can easily get anything in two clicks (ten tabs, 60 per tab without scrolling). Plus it has extra settings for tweaking images and text separately which comes in handy sometimes. I was vaguely aware of the zoom levels being remebered per site, but its been so long I wasn't sure. I might test it :)

**Comment by earthlng** [comment]
=> moved from [Issue #15](https://github.com/ghacksuserjs/ghacks-user.js/issues/15) ...

@ crssi commented on 21 Feb
Observation report:
- **`privacy.resistFingerprinting`**
Overall havent found any serious problems yet, but there is an "inconvenience".
I guess this "login.aliexpress.com" is a perfect site for testing... lol
When `privacy.resistFingerprinting = true` and together with NoScript, it raises ClearClick event in NoScript.
Funny, when `privacy.resistFingerprinting = false`, there are no "ClearClick" events in NoScript extension.

@Atavic commented on 21 Feb
@crssi
You can block _Clickjacking_ with uBlock: https://github.com/gorhill/uBlock/issues/892

I remember having many issues with ClearClick (haven't used NS for years).

@crssi commented on 21 Feb
@Atavic Thank you, but here we have two different approaches with two different results (uBlock vs NoScript).
uBlock blocks frames which is undesired action in this case.
I would guess that this is a feature/bug in NoScript. Will investigate further.

@crssi commented on 7 Mar
As @Thorin-Oakenpants already noticed, I also do not find any problems at this time when TRUE.
There are no form jerks and also NoScript doesn't fires false ClearClick events.

**Comment by Thorin-Oakenpants** [comment]
Note: https://bugzilla.mozilla.org/show_bug.cgi?id=1345322 . So as of FF55 it is no longer a hidden pref. Just making a note to remind us when we move to v55

**Comment by Thorin-Oakenpants** [comment]
Interesting: https://bugzilla.mozilla.org/show_bug.cgi?id=1333933#c6

> In my implementation, these prefs will be updated by nsRFPService, that it will observe the change of 'privacy.resistFingerprinting' and updates these prefs properly. Following is what it will do.
>
>  1. When flipping 'privacy.resistFingerprinting' to true, the service will store  original values under the pref branch 'privacy.resistFingerprinting.prefsCache.' if they have a user setting value.
>  2. Set all prefs with values that can prevent browser fingerprinting.
>  3. If someone modifies these prefs when 'privacy.resistFingerprinting' is on, the change will be effective. And the service will update cached pref values. So, the user setting value won't be affected during disabling fingerprinting resistance.
> 4. When flipping 'privacy.resistFingerprinting' to false, the service will try to restore these prefs with cached values. If cached value doesn't exist, the service will restore it back to the default value.
>
> In addition, the nsRFPService will change prefs when it is the first run if 'privacy.resistFingerprinting' is turned on by default.

**Comment by earthlng** [comment]
check http://browserspy.dk/screen.php - FF54 changed something again that results in slightly different "height" values (for me)

**Comment by Thorin-Oakenpants** [comment]
Mine didn't change one bit. I check it every day :) I'm on FT Deep Dark and some CTR tweaks

**Comment by earthlng** [comment]
weird. I must have resized the window by accident then - thanks, good to know

**Comment by Thorin-Oakenpants** [comment]
see https://www.ghacks.net/2017/06/15/firefox-54-new-preferences-and-ghacks-user-js-changes/#comment-4213419 . Also see https://bugzilla.mozilla.org/show_bug.cgi?id=1364261

suggest these two warnings - yes? no?
```
/*blah

 ** hide the contents of navigator.plugins and navigator.mimeTypes (FF50+)
   [WARNING] This will break Flash
   [1] https://bugzilla.mozilla.org/show_bug.cgi?id=1281963

 ** spoof timezone as UTC 0 (FF55+)
   [WARNING] This will fubar your timestamps in web-based gmail
   [1] https://bugzilla.mozilla.org/show_bug.cgi?id=1330890

***/
```

**Comment by Atavic** [comment]
A different UTC has no breakage, only drawback is that what has been recently posted - like this comment - appears as posted 9 hours ago, in my case.

**Comment by Thorin-Oakenpants** [comment]
@Atavic I didn;t say breakage .. I said fubar :) Picture different devices - say a gmail app, using the web interface at home with resistFP, using the web site on a work PC, and using a client - now do you understand that one of these is not the same as the others re timestamps

**Comment by Thorin-Oakenpants** [comment]
IDK. I brought this up because of that comment at ghacks. But TBH, at what point do I stop the madness and pandering to people who can't comprehend things. Hiding navigator.plugins breaks your only plugin - go figure! Spoofing your timezone will affect timestamps - go figure! The only reason I would put them in is the warning tags make it easier to troubleshoot.

TBH, in the case of email, i'm expecting that the header (I am not an email expert) timestamps would cause chronoclasms. Here's an example, and I think I have this right (here's a map for you - https://www.timeanddate.com/time/map/ ), and lets ignore DST (daylight savings time)

Lets say that I'm in Sydney, Australia (UTC +10) and earthling lives in Vancouver, Canada (UTC-7). Yup, we're part of the Five Eyes. 17 hrs difference. I'm always ahead of earthlng xD.

I send a gmail to earthlng with the topic "Final instructions" and I am NOT `spoofing as UTC 0`, this is Wed 2am my timezone. Earthlng's email client will says it was sent Tues 9am as per his timezone. An hour later I send earthlng a follow up email with the topic "Correction, don't use the last info I sent or the world will explode", but this time I do spoof the time zone (this is Wed 3am Sydney time, but Tuesday 5pm UTC 0 time)

So check out my "sent items"
Wed 2am - Final instructions
Tues 5pm - Correction

How does this affect eathlngs inbox?
Tue 9am - Final instructions (it converted UTC +10 back into this own timezone but removing 17 hrs)
Tue 3am - Correction (it converted UTC 0 back into his own timezone by removing 7 hrs)

This is doing my head in a little, but I think I have it right. Pretty sure earthlng now uses the wrong info and blows the earth up. Amirite?

**Comment by earthlng** [comment]
> suggest these two warnings - yes? no?

Yes, but not as [WARNING]'s IMO - it's good enough if "Flash" is mentioned in a [NOTE]. 
ie Problem with Flash? - search the file for Flash, duh! problem solved.

> Amirite?

In your example you assume that gmail uses your timezone instead of the one from the server. IDK, is that the case?
not to mention that afaik the packets sent out from your machine have a timestamp in it anyway.

**Comment by earthlng** [comment]
re: [WARNING]'s - IMO we should also replace all warning tags on prefs that are inactive

**Comment by Thorin-Oakenpants** [comment]
> In your example you assume that gmail uses your timezone instead of the one from the server.

I'm not an email expert, but why would any email use server time - a server could be anywhere in the world, but users would require all time synced to their time-zone. It makes no sense to me to use server time, if it did, then everyone would be like USA time on gmail. gmail of course could do things differently and convert time based on your locale, IDK. Not my area.

https://www.lifewire.com/how-to-understand-date-and-time-in-email-headers-1170524

First article I found. Can't be arsed looking for any more. Everyone would use UTC, it would stamp the time sent as your timezone, everyone else just coverts it to their time zone so between Sydney and Vancouver it would just deduct 17 hrs. I don't think servers have anything to do with what it.

Wondering how to write the [NOTE] for the UTC spoof. Yes, I'll use [NOTE] (although the js header section says to search for [warning] and [setup] to troubleshoot)

> re: [WARNING]'s - IMO we should also replace all warning tags on inactive items
- 1201, 1212, 1221, 1241, 2421, 2514, 2621, 2671, 2705
- Not sure about that. A warning is a warning. The description contents are not dictated by pref status. If a user changes the status of a pref, that's immaterial. Take 2514 - if they enable it, the warning holds true. If they disable it, are you suggesting they should edit the description and downgrade the warning to a note (and vice versa)? That makes no sense to me.

PS: Wondering how to write the [NOTE] for the UTC spoof - I don't think "fubar" cuts it

**Comment by Atavic** [comment]
@Thorin-Oakenpants 

>Amirite?

Yep. Timestamps start from your PC. 

On Twitter I got messages posted some hours in the future!

**Comment by earthlng** [comment]
re: email time - from that article:

> each of these **servers** finds the time to record the current time

> How Can I Convert the Date and Time to My Time Zone?

--- 

> Not sure about that. A warning is a warning. ... That makes no sense to me.

ok, nvm, was just an idea.

**Comment by Thorin-Oakenpants** [comment]
The warning/note about email was not about tracking, it was about screwing up your timestamps as a potential downside - if the hops/info between point `sent` and `received` are out of whack (in the header?) is immaterial. The visible date (sent/received) is what the end user sees, and how things are ordered - that was all.

This is why they want a pref to opt out of UTC spoof with resistFP (https://bugzilla.mozilla.org/show_bug.cgi?id=1364261). It was just an idea to maybe add some sort of warning. Maybe I should just drop it and lets see what comments/tickets arise after 55 lands

**Comment by Atavic** [comment]
If fubar means the time will be off for some hours, the warning is fine.

**Comment by earthlng** [comment]
```js
// user_pref("privacy.window.maxInnerWidth", 1366);
// user_pref("privacy.window.maxInnerHeight", 768);
```
FYI these values don't work. you can only set it to round hundreds. 800x600 anyone? or 1600x900

**Comment by Atavic** [comment]
[Window.innerWidth](https://developer.mozilla.org/en-US/docs/Web/API/Window/innerWidth):

> Width (in pixels) of the browser window viewport including, if rendered, the vertical scrollbar.

**Comment by Thorin-Oakenpants** [comment]
https://reviewboard.mozilla.org/r/118948/diff/6#index_header - rounded AFTER using the 2 prefs

@Atavic A GM script will not cover all methods to determine measurements (css, js, and more) of multiple items (screen, browser, innners etc). using the gm script will simply fuck everything up and raise your entropy

**Comment by earthlng** [comment]
> I'm pissed off about this (I know they want to reduce entropy, but they provided an override and then SHIT on it)

Yeah but the good thing is that **without RFP** the nightly I have currently in testing is always reporting my real screen resolution for all the window values, even if the window is not maximized. That's good because with RFP enabled you can resize the window to any adjusted size to spoof any common resolution. 
The bad thing about the current implementation of the round hundreds thingy is that sessionrestore doesn't remember/restore your window size. It always opens FF with 1000x1000 if RFP is enabled.
But if you set the maxInner prefs you can have it open with 1600x900 for example, which is also fairly common. But fe. 14**40**x900 will open as 14**00**x900!

**Comment by Thorin-Oakenpants** [comment]
Interesting comment - https://bugzilla.mozilla.org/show_bug.cgi?id=1369309#c37  .. wonder what YT collect all that for then (nudge nudge wink wink, say no more, say no more)

**Comment by earthlng** [comment]
Thanks for keeping us updated about all of this, Pants. Damn I wish they'd release 56 as an ESR version.

**Comment by Thorin-Oakenpants** [comment]
But they will .. it will be called ... `<drumbeat>` .. Firefox 52.4

**Comment by earthlng** [comment]
no that's not how it works. 52.4 will only get the security fixes from 56

**Comment by Thorin-Oakenpants** [comment]
do I really have to put `/s` at the end

**Comment by Thorin-Oakenpants** [comment]
You'd hate 56 anyway, because resource URL breaks, pureURL breaks, FIndBarTweak breaks .. a lot of addons are already breaking. So it's probably best it isn't 56->ESR

**Comment by earthlng** [comment]
> because resource URI breaks

I already noticed that in nightly - sucks. Do you know why it broke? Have you read anything about it?

**Comment by Thorin-Oakenpants** [comment]
I only knew because kaliostro told us in the addons sticky 2 or so months ago - https://notabug.org/desktopd/no-resource-uri-leak/issues - I watched the addon page and the bug and nada .. no comment from the dev .. it's dead Jim

**Comment by earthlng** [comment]
Oh, must have missed that. Thanks for the link - bookmarked ;)
It's good to know that it already broke with FF55, maybe I can figure out why it broke and if the dev doesn't show up anytime soon, get a fixed version signed and distribute it here until the official version gets fixed.

**Comment by earthlng** [comment]
The problem with the addon has something to do with e10s. I managed to fix it but it's now non-e10s.
Since I don't use e10s yet anyway I didn't feel like wasting any more time on this to try and fix it properly. At least I can keep using it in 55+.

**Comment by Thorin-Oakenpants** [comment]
If you use CTR (Classic Theme Restorer), it trims 4 pixels off your height when you start (may depend on some CTR settings re padding etc - I turned all those off to defaults to strip CTR down to the bare minimum in prep for 57). It also happens when you use the privacy.window.max* .. so my best guess is the resize happens before the add-on takes hold. Not sure if this is the case with userChrome.css or whatever it's called.

**Comment by Thorin-Oakenpants** [comment]
Hmmm .. I now have a spoofed timezone, but GitHub still nailed me for commit times

**Comment by share3141** [comment]
I have completely removed CTR - in fact all legacy extensions except NoScript, Ublock Origin, HTTPS Everywhere, and Umatrix.  
I do not have any *.CSS files.  
My issue appears to be similar to the above - some pixels are being trimmed off the top.
I have set the properties of FF (desktop icon) to open "maximized" and the menu bar is almost invisible it is up so high on the screen.
When I change the properties of FF to open "normal" and then maximize the screen everything looks ok.  
I know it has something to do with the entries in user.js (THANK YOU VERY MUCH) because I have built a new profile and left out the user.js file and no issues appear.  I also introduced the user.js gradually to the new profile by copying sections of my other profile (user.js) to try and narrow down the issue.  No luck - I can't seem to figure out where it breaks.  
The privacy.window.max* settings are commented out and do not appear in the pref.js.  It first appeared in FF 55 beta and now appears in FF 56 beta.  It's not fatal behavior but it bugs me!

**Comment by Thorin-Oakenpants** [comment]
> My issue appears to be similar to the above - some pixels are being trimmed off the top.

I do not understand. The pref privacy.resistFingerprinting (2599a) does a LOT of stuff. One of the new things it does in FF55+ is round the window (inner size) to 100's (width by 200's and height by 100s). The two prefs privacy.window.max.* (2699b) allow a sort of override - eg you could set them to 1600x900 .. whatever they are, the rounding will still happen, and it does this to fit within your screen. It takes into account all the non-inner window bits, such as toolbars etc

So when you say "some pixels are being trimmed off the top" do you mean that the browser's title bar is slightly off the top of your screen?

Here are three suggestions
1. Don't open "maximized". Open "normal" and grab the bottom right edge and resize the window so that it has plenty of room around it, and grab the title bar and drag the window down properly into the screen. I suspect that the browser is remembering your starting X,Y coordinates for the top left corner (which being maximized is causing it to be a little too high). How close firefox. Now re-open it. The reason I say to try this, is because mine remember EXACTLY where to start on the screen

2. Disable privacy.resistFingerprinting (both in the user.js and by resetting it in about:config. Restart your browser. This would prove it's this pref causing the issue, which by the sounds of it all, it is exactly this

**Comment by share3141** [comment]
Well isn't this special!  
I tried option #1 and no luck.  
Option #2 took care of the problem.  
One - thank you.  
Two - what are the consequences of not having  privacy.resistFingerprinting set to "True"?

**Comment by share3141** [comment]
I have my task bar at the top...(I'm a very old former Apple user and never got out of the habit).  I'm trying - I dual boot with Linux and I haven't moved the task bar to the top.

**Comment by Thorin-Oakenpants** [comment]
@share3141

Option 3. Use privacy.resistFingerprinting, also set the two prefs for size to be small enough for your screen. Eg if you are on a 1920x1080 screen, set the browser to open as a nice small 1000x600

ie in the user.js edit these two lines in 2699b (remove the leading `//` chars and change the values
```js
user_pref("privacy.window.maxInnerWidth", 1000); // (hidden pref)
user_pref("privacy.window.maxInnerHeight", 600); // (hidden pref)
```

Resize Firefox so easily all within your screen, and then close it. Now open it. Did it open as a smaller size at around 1000x600? Did it position all within the screen?

Interesting you say the taskbar is at the top - this may be the factor. Maybe you could also do another test with the taskbar at the bottom or side - just unlock it and drag it for the test.

---
Consequences of not using the privacy.resistFingerprinting is you don't get *some* of the things it covers. I say some, because many of the things this pref does is merely ENFORCING what other prefs already do by overriding them (sometimes in a different way).

See this partial list (from the user.js in section 2699)
```js
/*
 ** 1281949 - spoof screen orientation (FF50+)
 ** 1281963 - hide the contents of navigator.plugins and navigator.mimeTypes (FF50+)
 ** 1330890 - spoof timezone as UTC 0 (FF55+)
 ** 1360039 - spoof navigator.hardwareConcurrency as 2 (also see 2514) (FF55+)
      This spoof *shouldn't* affect core chrome/Firefox performance
 ** 1217238 - reduce precision of time exposed by javascript (FF55+)
 ** 1369303 - spoof/disable performance API (see 2410-deprecated, 2411, 2412) (FF56+)
 ** 1333651 - spoof Navigator API (see section 2697) (FF56+)
      The version number will be rounded down to the nearest multiple of 10
 ** 1369319 - disable device sensor API (see 2512) (FF56+)
 ** 1369357 - disable site specific zoom (see 2515) (FF56+)
 ** 1337161 - disable gamepad API (see 2501) (FF56+)
 ** 1372072 - spoof network information API as "unknown" (see 2503) (FF56+)
 ** 1372069 - disable geolocation API (see 0201) (FF56+)
 ** 1333641 - disable WebSpeech API (see 2021) (FF56+)
***/
```

See the ones that have a (see xxxx), well those ones can be handled without privacy.resistFingerprinting. Example "disable device sensor API". Most of these are yet to come in FF56+. An example of something there is NO other pref for is spoofing your time zone as UTC (which landed in FF55).

**Comment by share3141** [comment]
Option 2 seems to resolve the issue but of course creates another one - privacy.resistingFingerprinting isn't set.  I did move the taskbar to the bottom and did option #1 and #3 again.  No change.  Not fatal - just an extra click.  At least I know know what is causing it so that makes me feel better.  Thanks for your time.

**Comment by Thorin-Oakenpants** [comment]
@share3141 I encourage you to let the Mozilla guys know. When this first landed there were all sorts of weird dialog sizes in Nightly, and subsequent bugs filled. File a bug and provide a screenshot of the "cut off" bit - unless they know, they can't fix it.

**Comment by earthlng** [comment]
With privacy.resistingFingerprinting=true you're not supposed to use a maximized window

**Comment by Thorin-Oakenpants** [comment]
@earthlng While I cannot tightly control the testing, he still had the issue with opening it in normal mode (and even after it was resized and positioned nicely manually and closed to preserve the x,y coords).

**Comment by share3141** [comment]
I was unaware that with privacy.resistingFingerprinting=true you're not supposed to use a maximized window.  Where is that info?  So much to learn!

**Comment by Thorin-Oakenpants** [comment]
Sorry @share3141 - that should be painfully aware. If it is trying to set your `inner windows` to round hundreds, it cannot conform to a maximized state because it has to resize the `browser window` - in a maximized state the `browser window` cannot be resized

**Comment by Theemim** [comment]
My feeling is that it is a mistake for Mozilla to remove, or otherwise not support, individual prefs which can be used to control the same things that resistFingerprinting can. Because there are cases where enabling things with fingerprinting potential makes sense and could (should) be controllable on a per site basis. For example, fingerprinting is used by some banks and other orgs to recognize the devices you log in from.  Without granular control, a user running into any problems at any important site (even one that may not be performing or fingerprinting) will disable resistFingerprinting and lose all protections?  I mean, if Mozilla continues down this pat?

**Comment by Thorin-Oakenpants** [comment]
That's what profiles are for. There will also be an option the UI at some point, but that's not very handy. But the recognise device is handy for a lot a services. resistFP is optional, it is not enforced. I agree that some of these anti-FPing measures would ALSO be nice under another pref.

I do however agree that *enforcing* what resistFP does is best to create the biggest set of FF users with identical values. However, a "whitelist" feature would be handy. Someone should suggest it in a ticket. IN the UI, to the right of he check box (when it lands), add a button [...Exceptions]

**Comment by 2glops** [comment]
In 2699, the advise to use window.resizeTo should not be removed ?
With resistFingerprinting set to true, using window.resizeTo works, but only for the session. The setting is not persistent after restart.
2699a give the needed information for resizing the browser window.

**Comment by Thorin-Oakenpants** [comment]
Nah. The info in the section header should stay 1) good info [I have to fix mine in FF55 on every start because CTR throws it out by 4px height] 2) ESR users 3) The 2699b explains what happens after 55+. I do not wish to duplicate info in 2699b, plus it does not belong there as people _shouldn't be resizing FF55+_ (but can tweak their new window sizes somewhat with the two `privacy.window.maxInner` prefs)

`set new window sizes to round to hundreds` .. let me emphasize that a little ... "set NEW WINDOW sizes...". Clearly when you start FF you create a new window. Any manually resizing is not meant to be persistent.

**Comment by ArchangeGabriel** [comment]
Well, I use Firefox in maximized windows all the time, so this might not be for me…

On Panopticlick, with resistFingerprinting:
Screen Size and Color Depth | 14.76 | 27658.0 | 1920x944x24
Time Zone | 3.91 | 15.07 | 0
Platform | 1.28 | 2.43 | Win32
User Agent | 10.11 | 1104.11 | Mozilla/5.0 (Windows NT 6.1; rv:50.0) Gecko/20100101 Firefox/50.0

Without:
Screen Size and Color Depth | 2.3 | 4.93 | 1920x1080x24
Time Zone | 2.08 | 4.22 | -120
Platform | 3.18 | 9.07 | Linux x86_64
User Agent | 13.87 | 14950.38 | Mozilla/5.0 (X11; Linux x86_64; rv:57.0) Gecko/20100101 Firefox/57.0

And I’m not fond of switching my platform to Win32, even if it’s far less identifying obviously. So I would prefer a way to just change the UA to latest release instead of Nightly. I guess I can only go for manually editing it?

**Comment by Thorin-Oakenpants** [comment]
Fingerprinting test results (the entropy, not the returned values) are batshit bullshit crazy - see #195 

Why the hangup over spoofing your OS? Either get on board with a set of enforced values via resistFP, or absolutely almost certainly get it wrong and stick out like a dog's balls

resistFP is likely (hopefully) to become part of PB mode at some stage, after it is implemented as an option in the preferences UI (where it will gain more traction as well). Windows rounded to hundreds will fail a tad if you are tweaking the UI (chrome.css or whatever it is, or eg at the moment with Classic Theme Restorer), but it is not designed to work with maximized windows for all the reasons you already know. But if you always work maximized, then big deal, I fail to see how this affects you - open the browser, click maximized (or am I missing something?) Your choice to continue having fairly unique inner window measurements

1920x1080x24 <- this is your SCREEN measurement, resistFP standardizes screen, available screen, browser, and inner windows all as the same (they use inner window so it does not FUBAR web content) = way less measurement points and variations caused by them. It also covers other screen things such as orientation etc

**Comment by ArchangeGabriel** [comment]
My problem with the OS is that this would be lying a lot and as you said in #195, very likely to create a mismatch.

Regarding windows size, I now understand thanks to your edit. It’s annoying to have to ask Firefox to go maximized each time I launch it though…

In the end I get that for now, resistFP makes me an outlier because of the mismatch it might causes. But once more people will use it (thanks to being more easily discoverable and the default in PB), it will makes me one of the many person with the same mismatches, while other settings not subjects to mismatch will be spoofed to a common value.

**Comment by Thorin-Oakenpants** [comment]
Unfortunately detecting OS is easy (scrollbars and CSS line heights differ between OS) but there are plans to patch those. All you can do is trust in the Tor Uplift people. Winging it with your own decisions/values will likely sink you

> resistFP makes me an outlier ..

Compared to doing no spoofing, you will stand out a little (especially if not on Windows and they detect you're on OS). Compared to spoofing it yourself, resistFP makes you part of a much larger group. As I said elsewhere, imagine 1% uptake - you're already part of a million users. Wait til it gets 10% (and some holes closed) :)

At the end of the day FPing is fun (for me), and certainly reducing the "attack" surface, spoofing, blocking, etc is always good. Make it HARD for the bastards. That said - I probably use around 200 sites with JS .. I block all XSS & JSS by default (the other 5000+ sites I visit over a year it is all blocked). I think FP'ing gets way too much hype - kill JS and you kill 95% of your problems (not to be confused with tracking). Obviously there is more (server side fingerprinting eg timing, SSL ticket sessions, etc and other non-JS methods), but JS accounts for most of it. Don't sweat it too much.

You're on 57: Didn't realize that they had implemented the warning when going maximized yet.

**Comment by ArchangeGabriel** [comment]
So we agree.

Yes, I also block JS on most sites. Well, I used to, but NoScript is not working anymore on Nightly (they killed non-WE addons some days ago)…

Sorry if I was unclear, but I don’t remember having mentioned a warning. And I’ve seen none. If you refer to:
> It’s annoying to have to ask Firefox to go maximized each time I launch it though…
Then what I was saying is that with resistFP on, Firefox open on a fixed window size which is not maximized, so I have to click on the maximize button at each launch.

**Comment by Thorin-Oakenpants** [comment]
I misread it, soz. They are going to implement a warning - ~~see first post~~  hmmm, could have sworn there was a ticket somewhere for this, maybe I got confused with the proposed warning for changing language.

**Comment by rekixex** [comment]
@earthlng 

> With privacy.resistingFingerprinting=true you're not supposed to use a maximized window

Interesting, I missed that. So, users who _always_ open Firefox in a _maximized_ window should keep privacy.resistFingerprinting pref on **false**? What would you do?

**Comment by rekixex** [comment]
@earthlng 

> With privacy.resistingFingerprinting=true you're not supposed to use a maximized window

Interesting, I wasn't sure about that. So, users who _always_ use Firefox _maximized_ should keep privacy.resistFingerprinting pref on **false**? What would you do?

**Comment by Thorin-Oakenpants** [comment]
> So, users who always use Firefox maximized should keep privacy.resistFingerprinting pref on false

No. They should keep it on true, because it does so much _more_ than just spoof screen-res/various-window-measurements. Maximizing is not forbidden, it just destroys your actual inner window measurements which are used for the spoof values. If a user wants to maximize, then they will probably become reasonably unique (I say **reasonably** because the subset of maximized values on various screen res with various browser elements/tweaks is going to be small'ish - maybe a couple hundred variations - not as small as the enforced rounded hundreds, but smaller than 1 pixel variation combos)

The silly thing with FP'ng tho, is get _one_ (or two) thing(s) wrong and all the benefits go out the window ( see https://bugzilla.mozilla.org/show_bug.cgi?id=1333933#c5 ).

> Partial fingerprinting protection is not very useful

That said, there are things not patched yet and too many attack vectors. We're already unique (in worse case scenarios, eg JS on). However, choosing to ignore a patched attack vector is not an answer. Adapt is what I would say.

**Comment by earthlng** [comment]
> So, users who always use Firefox maximized should keep privacy.resistFingerprinting pref on false? What would you do?

I would keep privacy.resistFingerprinting set to true. Maximized is still far better than a non-maximized window that doesn't match any common screen resolution.

https://earthlng.github.io/testpages/screen_information.html

I took the screen information page from browserspy.dk, removed a few things and added a couple other ones instead. You can play with it a bit, with resistFingerprinting on and off, and with maximized and non-maximized windows and see what a site could potentially gather from you.
Making additional toolbars always visible (FF Menu->View->Toolbars) changes the values as well.
With a non-maximized window you can adjust for that and effectively hide your real screen resolution and whether you have the menubar and/or bookmarks toolbar (and/or sidebars for that matter) always visible.

In FF55+ you can set those 2 new prefs (maxInner something) and it will automatically make those adjustments for you.

**Comment by rekixex** [comment]
@Thorin-Oakenpants @earthlng Thank you for the clarification.

> Maximizing is not forbidden, it just destroys your actual inner window measurements which are used for the spoof values. If a user wants to maximize, then they will probably become reasonably unique (I say reasonably because the subset of maximized values on various screen res with various browser elements/tweaks is going to be small'ish - maybe a couple hundred variations - not as small as the enforced rounded hundreds, but smaller than 1 pixel variation combos)

This is especially true for Linux. Various desktop environments have different panel sizes in pixels, which directly affects available screen height/width (see https://browserleaks.com/javascript). On Windows, with the default appearance settings (i.e. taskbar size), the default Firefox toolbars and a common screen resolution, the entropy shouldn't be much higher.

> In FF55+ you can set those 2 new prefs (maxInner something) and it will automatically make those adjustments for you.

I did some testing and it seems that `privacy.window.maxInnerWidth` and `privacy.window.maxInnerHeight` can't be set to acceptable values for a reasonably wide/high window on a 1366x768 screen resolution (which is the most common one).

Which browser elements affect screen resolution? Only the menu bar and the bookmarks toolbar? Does customization matter or only toolbar state? What about customization of the **navigation toolbar**?

**Comment by Thorin-Oakenpants** [comment]
< Which browser elements affect screen resolution

Everything. Showing and hiding any elements such as statusbar, sidebar, findbar, toolbar(s) eg bookmark toolbar, menubar locationbar, and yes, scrollbars.  Oh yeah .. and ZOOM as well .. and the actual browser itself - paddings, etc (eg changes caused by themes, userchrome.css and tweaks from CTR etc all affect the inner window)

**Comment by earthlng** [comment]
> Aug-11: disable media statistics resolved fixed FF57+, reworded as spoof to 0

they don't spoof to 0, see part 3 of that bug @ https://hg.mozilla.org/mozilla-central/rev/9afd71d7438c

**Comment by earthlng** [comment]
> disable gamepad API

The same as I wrote in my last comment also applies for FF56 + `dom.gamepad.enabled`.
Setting `dom.gamepad.enabled` to false disables the API and can be detected.
Starting with FF56, resistFingerprinting will provide a better solution for this ie not disabling the API and instead "hiding" any connected gamepads from websites:
1. navigator.getGamepads() will always return an empty array
2. gamepadconnected and gamepaddisconnected events will not be fired

ps. "disable gamepad API" is incorrect because that's not what the patch does.

**Comment by earthlng** [comment]
https://github.com/ghacksuserjs/ghacks-user.js/issues/7#issuecomment-320970736
> See the ones that have a (see xxxx), well those ones can be handled without privacy.resistFingerprinting.

we should look at all those as well.

**Comment by Thorin-Oakenpants** [comment]
When it lands, when the ESR + FF branches merge and resist fingerprinting can stay enabled applies to everyone, a bunch of other prefs can be made inactive (in the user.js) - until then we leave them on. Some of the resistFP measures are more elegant that the other prefs which eg block rather than spoof.

When I put (see xxxx) I did not mean that those prefs do the same thing. Some approaches are different (spoof, block, whitelist etc).

I was following what they did in the tickets, not necessarily the bug title - without looking it up, they explicitly said return 0 instead of block or whatever, way down the ticket, at the end or whatever. I've tried to make the description mean exactly what they did - sometimes they keep and chopping and changing within that ticket

We should probably go thru the current landed resistFP tickets and confirm exactly what the code does

**Comment by Thorin-Oakenpants** [comment]
^^ie due to the morphing nature of their solutions, I have not always picked up the final changes - so lets create a ToDo and check em all

PS: UA version is going to be spoofed to ESR rather than rounding down to multiples of 10

**Comment by rekixex** [comment]
@Thorin-Oakenpants 

> > Which browser elements affect screen resolution
> 
> Everything. Showing and hiding any elements such as statusbar, sidebar, findbar, toolbar(s) eg bookmark toolbar, menubar locationbar, and yes, scrollbars. Oh yeah .. and ZOOM as well .. and the actual browser itself - paddings, etc (eg changes caused by themes, userchrome.css and tweaks from CTR etc all affect the inner window)

Even the _customization_ of the **navigation toolbar**? Many extensions place icons there and I guess users often add/remove/rearrange navigation toolbar items (like the subscribe button which is hidden by default).

**Comment by earthlng** [comment]
No. Simply adding or removing extension icons doesn't change the height of the toolbar.

**Comment by ArchangeGabriel** [comment]
@earthlng It has not been true in the past. I remember the Ghostery icon being twice the normal size for quite some time, and thus increasing the toolbar height. I think it was some kind of HiDPI issue.

**Comment by earthlng** [comment]
Oh okay, thanks for clarifying.

**Comment by Thorin-Oakenpants** [comment]
> Even the customization of the navigation toolbar?

i am not sure of the exact order of things.. but BEFORE the browser is rendered, it has decided your measurements. Any change thereafter can affect the inner window, because it is not the overall browser width and height that change.

It is at this point, AFTER the RFP code, that _unexpected_ things can happen, such as: HTTPS Everywhere had a release with a GIANT icon (ewww to Ghostery, so don't know about that one), or Classic Theme Restorer kicks in and tweaks padding to various UI elements. I do not know when userchrome.css kicks in - maybe someone can test that.

PS: IMO they could make the Findbar a doorhanger so it doesn't influence measurements, but then again, all RFP users are in the same boat, and the occasional obfuscation probably helps

**Comment by Thorin-Oakenpants** [comment]
FYI: https://www.reddit.com/r/firefox/comments/6xywm9/weird_time_sync_issue/

^^ f**kbook, twitter, etc

Strangely, github gets my timezone some other way (and it's not a user account setting) - I can see it in the commit timelines (I am as far ahead of UTC 0 as possible, so the diff sticks out like a dog's balls). Maybe it's using IP=>geo ? It gets my timezone correct btw.

:warning: 10MB PNG file
https://upload.wikimedia.org/wikipedia/commons/e/e8/Standard_World_Time_Zones.png

Yup, that's me .. waving at you from Петропа́вловск-Камча́тский .. send me a postcard

**Comment by earthlng** [comment]
> * Sep-05: 1396468 added to pending - this is a fix to one of those navigator.* leaks I have bitched about for years

Less bitching and more testing + reporting would help ;)

**Comment by Thorin-Oakenpants** [comment]
* 1382533 add fingerprinting resistance to Presentation API
   *  don't expose local IP Addresses via mDNS (Multicast DNS)

I'm not sure on how to word this for the user.js. It has the same effect as disabling Presentation API AFAIK by blocking request and receiver

**Comment by Thorin-Oakenpants** [comment]
FYI: https://bugzilla.mozilla.org/show_bug.cgi?id=1333641#c8  `1333641 disable WebSpeech API` will be renamed appropriately in a commit shortly. It does not disable the API at all

> we should make the SpeechSynthesis.getVoices() return an empty list, block 'onvoiceschanged' event and make SpeechSynthesis.speak() reports a 'onerror' event

And that's what they did in the end. Also from this conversation they created [1233846](https://bugzilla.mozilla.org/show_bug.cgi?id=1233846) as a followup to add further anti-FPing to the API

**Comment by Thorin-Oakenpants** [comment]
FYI: https://bugzilla.mozilla.org/show_bug.cgi?id=1404017 - enable RFP for PB mode only. Same as https://bugzilla.mozilla.org/show_bug.cgi?id=1397624 - enable FPI for PB mode only.

Note this is proposed only but highly likely IMO
> 0: Fingerprinting Resistance off
> 1: Fingerprinting Resistance on in all windows
> 2: Fingerprinting Resistance on in Private-Browsing windows only

If these both go the route of changing the prefs from boolean to integer, I am not sure what will happen with ESR. If the change is not made in ESR, then I guess we can replicate the pref as boolean in the deprecated section

**Comment by Thorin-Oakenpants** [comment]
FYI: Starting in FF58 .. WE have access to this pref - https://bugzilla.mozilla.org/show_bug.cgi?id=1397611

**Comment by Thorin-Oakenpants** [comment]
heh - https://www.youtube.com/watch?v=TQxuuFTgz7M .. I don't like it, controlling the viewport

**Comment by Thorin-Oakenpants** [comment]
FYI: https://hardware.metrics.mozilla.com/ - Oct 22 2017 (latest stats) .. Browser Architecture .. WOW, that's a massive swing there between 32 & 64bit (45% vs 55% - was 66% vs 33% and 81% vs 18% in previous weeks) now the DL site & updates push 64bit where it can

The ticket was ahead of it's time: https://bugzilla.mozilla.org/show_bug.cgi?id=1383495

**Comment by Thorin-Oakenpants** [comment]
** 1372073 - spoof/block various media device fingerprints (FF59+)

^^ not moved to resolved yet in first post yet. the ticket says the component is "WebRTC: Audio/Video" but I think this **may** covers more than WebRTC - I just want to get the wording right. Don't get me wrong, come FF59, we will be choosing to keep WebRTC disabled. Also note there is a new permissions UI for camera/mic with site exceptions in 57+. Earthlng or someone, can you give me some feedback on this - https://bugzilla.mozilla.org/show_bug.cgi?id=1372073

**Comment by earthlng** [comment]
bump again when we get close to 59

**Comment by Thorin-Oakenpants** [comment]
^^ Awww man .. that ruins my workflow/methodology. The ticket just closed -> time to edit first post and add to js -> no longer have to deal with it, no need for reminders. Diffs for releases is the other trigger point, if I remember.

**Comment by Atavic** [comment]
MediaDevices is a different API but a tool like [WebRTC JavaScript library](http://www.rtcmulticonnection.org/) implements the [Mediadevices API](https://www.webrtc-experiment.com/RTCMultiConnection/getMediaDevices.html) into WebRTC.

**Comment by Thorin-Oakenpants** [comment]
reminder: follow up on adding `privacy.spoof_english` 1039069 as 4504 - can't access DXR right now, not sure if this should be documented in user.js, if so probably inactive as it seems to be controlled internally when users make language changes and/or flip RFP (by checking on first http connect)

**Comment by Thorin-Oakenpants** [comment]
FYI: if you were wondering about `1422862` `Canvas: OffscreenCanvas doesn't respect Canvas Permission Prompt` under the followups section ... OffscreenCanvas is not even enabled in FF - it is behind a pref which is default false anyway

```js
/* 2028: disable offscreen canvas
 * [1] https://developer.mozilla.org/docs/Web/API/OffscreenCanvas ***/
user_pref("gfx.offscreencanvas.enabled", false);
```

**Comment by jawz101** [comment]
One thing that prevents me from using fingerprinting on Firefox for Android is the screen resolution thing.  It makes no sense to blow a webpage up to full resolution when there's no way my phone will display it properly.

**Comment by Thorin-Oakenpants** [comment]
@jawz101 The UA spoof I believe causes desktop pages to be served. FF59+ should fix this as it stops trying to hide Android as the platform. I actually wouldn't use this on mobile yet

**Comment by aesthicc** [comment]
Would it be possible to include a version of the user.js that has privacy.resistFingerprinting set to false? I did the setup in my user-overrides.js, I just think i might miss changes made in the future. Just an idea for ease of use.

**Comment by Thorin-Oakenpants** [comment]
Not interested in creating an app - too much work.

As for the privacy.resistFingerprinting (RFP), it is just ONE pref (ok, and some for windows sizes if you want that, and that mozAddonManager one - but they're all inactive), so if you do not want to use RFP, then its one line in your overrides section. That's about as simple as it gets.

```js
/* My Overrides ***/
user_pref("privacy.resistFingerprinting", false); // 4501: i am not ready for RFP
// reminder on updates to re-enable the 4600 section **IMPORTANT**
```

I know there are a ton of things that are covered under RFP, and more to come, but its still one pref. Nothing to miss except keeping up to date with what RFP covers - which is what this sticky issue is for, as well as the ever-growing list of items in the user.js 4500 section header

Also be aware that if you do not want RFP, then you can also flip section 4600 on by changing one character (eg to get back some timing attack mitigations, etc), although the whole section is inactive BUT you would have already applied it (assuming it the section was previously enabled), except anything new moved there sine your last update (but these also would have been in their original state prior to being moved there).

**Comment by Thorin-Oakenpants** [comment]
@overdodactyl  the bug you want to tell these people here ( https://www.reddit.com/r/firefox/comments/80yqnx/google_apps_privacyresistfingerprinting_problems/ ) is this one

[1433592](https://bugzilla.mozilla.org/show_bug.cgi?id=1433592) Keyboard: don't spoof/suppress CTRL key (regression from 1222285 FF59+)

**Comment by overdodactyl** [comment]
Thanks Pants! Just passed along the info :)

**Comment by Thorin-Oakenpants** [comment]
@overdodactyl Why can people not learn to read. Marked solved as follows

[OP on reddit](https://www.reddit.com/r/firefox/comments/80yqnx/google_apps_privacyresistfingerprinting_problems/)
> EDIT: Exceptions on a per-site basis are not currently possible, due to this bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1433592). However, RFP will be opened to the Privacy API, so extensions like uBO should be able to utilize it.

The bug linked is about reverting the breakage that the CTRL key causes due to the keyboard FP mitigation in 1222285. It has nothing to do with the Privacy API & extensions .. ***which is already available in 58*** (its right there in the first post in this issue, if you want the bugzilla). The privacy API is not a solution anyway, because it's a global switch (what if you load a dozen sites at once or session restore a 100 tabs etc). The solution for RFP (which is not ready yet for the masses) is to reduce some of the breakage, patch holes, test like crazy, maybe do some studies, add to PB mode as a trial .. and maybe add a UI setting.

**Comment by overdodactyl** [comment]
Hmm ya, I either wasn't clear with what I said or it was misinterpreted a bit.  

After linking the ctrl bug I proposed possibly using a different profile for google apps if the bug was preventing them from keeping RFP on.  The mention of the privacy API being able to use it was more of a suggestion towards, in the future, uBO and other extensions might be able to quickly turn it on/off. 

I agree it's not a solution, but if someone's not willing to keep RFP on because of a bug like this, it's a reasonable alternative in an isolated profile to be able to quickly toggle it on/off as needed.

**Comment by Thorin-Oakenpants** [comment]
Not your fault, people lack verbal reasoning. I always tell my partner, that if she/he does something (eg the dishes, feeds the cat, whatever), that I **might** do something awesome (eg take us out for dinner, or go on a holiday - depends on the job I want him/her to do so I don't have to do it). And my partner does it, every single time ...

Then comes the inevitable question and wailing and tantrums and sad face down the track when my partner says "But you said you **would** do xyz" and I have to explain to them that I said **might** which is not the same as **will** and.. then I get no sex for a day until I promise I **might** do something in return .... *sigh*

**Comment by Thorin-Oakenpants** [comment]
@overdodactyl if its not there by the time you read this https://www.reddit.com/r/firefox/comments/82kyv8/firefox_screenshot_saving_blank_images_with/ for please let them know it is https://bugzilla.mozilla.org/show_bug.cgi?id=1412961

**Comment by overdodactyl** [comment]
Just did, thanks :)

**Comment by Thorin-Oakenpants** [comment]
FYI: https://fruct.org/publications/abstract21/files/And.pdf .. am reading it now. I like Arthur's (& TBB's) approach of lowering entropy, but they're taking each item in the Tor Uplift on its own merits - sometimes blocking, sometimes spoofing, and sometimes reducing accuracy.

It'll be interesting to see what they do with the OS in UA because lowest hanging fruit would suggest they might as well stay spoofing all as Windows and *if* some site wants to exploit the OS as an inconsistency, then it would be *no worse* than if they didn't lie about it in the first place. They other factor is site breakage due to OS.

**Comment by Thorin-Oakenpants** [comment]
Interesting roapmap: https://wiki.mozilla.org/Firefox/Roadmap

Webrender: aimed at FF64
Privacy Stuff: aimed at end of Q3 - so probably FF63 will get a UI for RFP

**Comment by Thorin-Oakenpants** [comment]
Yay .. ctrl key fix pushed in 58.0.2 - https://bugzilla.mozilla.org/show_bug.cgi?id=1433592  .. its been dodgy doing things on github to say the least - totally broken ctrl keys in edit

**Comment by earthlng** [comment]
> Mar 20: 1433592 (don't spoof ctrl key) changed to FF60 (the 59 patch was backed out)

change that back to 59.0.2, I guess?

**Comment by Thorin-Oakenpants** [comment]
I made a note on a sticky and will deal to it later both in OP and js .. I need to go practice my dragon slaying skills

**Comment by Thorin-Oakenpants** [comment]
Oh boy, this will be interesting ( https://bugzilla.mozilla.org/show_bug.cgi?id=1407366 ) .. they're going to move to 128'sx100's (I don't get it?)

I just got a new primary monitor (2560x1440) and open at 1600x900 (it's actually a pixel out due to the bookmarks toolbar which they still haven't fixed) and have two presets in my `Window Resizer` extension to get 1920x1080 and 1600x900

Time will tell, that demo is not quite what they were following I don't think, as they're talking about margins/padding in the inner view. Fuck this is gonna get ugly I reckon.

They would be better off just limiting the browser to 10 different preset sizes or something. I think this might be THE biggest turn-off for users up-taking RFP to be honest. This 128x100 is BS - how many RFP users are going to be eg 1536x1000 or 1664x1100 - answer: fucking NONE. Not a fan of this at all

**Comment by earthlng** [comment]
not sure what you mean by "affect" but I just tested with a japanese FF63 and the string in the brackets is in japanese characters even with RFP enabled.
They have a test for this but that test most likely only ever runs on en-US versions and therefore they didn't notice the problem.

**Comment by Thorin-Oakenpants** [comment]
It won't affect TBBv8 because it's a very recent change/regression. [Comment 3](https://bugzilla.mozilla.org/show_bug.cgi?id=1501875#c3) lists options. If they don't suppress leaking locale via this, then RFP will need to do so. At the moment RFP advises you should be using en-US (warns you otherwise). I'm not savy enough to work this stuff out, but I swear locale is leaked in more places (eg via formatting)

**Comment by earthlng** [comment]
correction!
during testing a prompt came up at one point asking me this:
> 言語設定を英語に変更すると、個人情報の特定がより難しくなりプライバシーが強化されます。英語版のウェブページを表示しますか？

Obviously I had no idea what that means and clicked the button which I assumed would be "cancel" because I thought it's something Activity-Stream or telemetry related.

Turns out this translates to
> Changing the language setting to English will make it more difficult to identify your personal information and enhance your privacy. Do you want to view the English version of the webpage?

and was triggered by me enabling RFP. Clicking the other non-cancel button and testing again shows that RFP then spoofs the timezone correctly in english as `(Coordinated Universal Time)`

**Comment by Thorin-Oakenpants** [comment]
FYI ... can't believe this wasn't on the radar already - [1507879](https://bugzilla.mozilla.org/show_bug.cgi?id=1507879) getClientRects .. just another ticket instigated by yours truly

**Comment by Thorin-Oakenpants** [comment]
FYI: https://bugzilla.mozilla.org/show_bug.cgi?id=1407366 landed in Nightly behind extra prefs

**Comment by Mikaela** [comment]
I am not sure if this is relevant here, but i have heard suspicion that enabling `privacy.resistFingerprinting` leads to higher amount of Google Captchas with them being more difficult. For me this only had the result that I started closing pages whenever they prompted me to fill a Google Captcha.

**Comment by Thorin-Oakenpants** [comment]
If I do too many google searches in a short space of time...

some background: for some downtime, I play trivia bots on IRC. I'm good. I go for about 3 hours, blast music, drink beers, wear no pants. I'm fast. I nail almost every question. Q's are 15 seconds apart. 50% of questions, I have no idea or am impatient with the clues (which are at 15, 30 and 45 seconds, the question times out at 1 minute), so I use my google-fu. searching for key words and strings etc. And I'm fast. Depends on the question, but I can copy content from my IRC client when the question is asked, paste it into firefox's search box, get my two-column dark-mode sanitized key-word-colorized auto-paginating google results, spot the answer visually, auto-copy the answer when I select the text, and paste it in as my answer in IRC and get the win .. **in about 10 seconds**

And its 15 seconds between q's. If I do three or four of those google searches in a row, then google thinks I'm a bot and nothing I can do will convince them otherwise on Firefox. I can prove I'm not a bot ob Chrome, but Firefox still can't use google. Chrome can. I can have the two open side by side when this happens, so it's not tied to my IP. And in effect, non chrome browsers are blocked by google for 12 hours. I actually mitigate this now by being careful with how fast I do my google-fu (and I have multiple custom search engines - one for wiki, one for IMDb, etc). Haven't been accused of being a bot now for about a year!! Yay!! I'm a human .. love me.

This is an example of what google do to shit on everything that is not theirs. They build in artificial delay timers if you're not using chrome. They display **really compressed** base64 thumbnails in image searches on Firefox, but on chrome those thumbnails are nice and clean. The list is endless.

When it comes to reCaptcha, I just refuse. I don't even think I can get it to work in my Firefox if I even tried to make it work. I actually run and maintain an Alexa top 10 thousand US site, and I got the site devs to exclude admins and moderators from captcha, that's how much I fucking hate it.

If I really have to use it, which is rare, I just use a secondary browser, which is auto sanitized. And yes, if you're not on chrome (which also leverages if you're logged in to your google account), then you get really shitty pics (but there's a reason for that), get asked multiple times, etc. It's a real PITA. The reason the pics have changed over time are because they need to keep making it harder and harder for the AI to keep learning - hence grainy pics, signs in foreign languages, vague questions over traffic lights (does it include those not facing the camera, does it include the pole), overlapping squares with bit pieces of the object, etc.

As for reCaptcha, there are probably a bunch of reasons. Tor uses get the worst of it I think. First of all, IP plays a role: VPNs, Tor exit nodes, etc. Then comes the browser: chrome gets special treatment. Logged into your google account (especially on chrome) will help. Then comes other factors - such as disabling SSL session ids & blocking cookies etc (all of which raise flags at google's end), RFP (this does actually break reCaptcha v3 - see https://bugzilla.mozilla.org/show_bug.cgi?id=1503872 ... and so on.

But I'm not a reCaptcha expert, and I don't plan on becoming one. I just cut it out of my life

**Comment by Thorin-Oakenpants** [comment]
OMG - quoted as a reference in Wikipedia - https://en.wikipedia.org/wiki/Device_fingerprint#cite_note-10

**Comment by earthlng** [comment]
FF70: [Change `outputLatency` and `getOutputTimestamp` when `resistFingerPrinting` is enabled](https://bugzilla.mozilla.org/show_bug.cgi?id=1564422)

**Comment by Thorin-Oakenpants** [comment]
yeah, sorry: been busy. Added to the list above. Did you want this added to the list in the user.js?

FYI: https://bugzilla.mozilla.org/show_bug.cgi?id=1324545 <-- AudioContext `getOutputTimestamp()` was added in FF70

**Comment by Thorin-Oakenpants** [comment]
not maintaining this anymore - it's already behind: info is in the user.js
